File: algorithm/src/main/java/com/qibenyu/algorithm/Tree.java
package com.qibenyu.algorithm;

import java.util.ArrayList;
import java.util.List;
import java.util.Stack;

public class Tree {

    static class TreeNode {

        int val;
        TreeNode left;
        TreeNode right;

        public TreeNode(int val) {
            this.val = val;
        }
    }

    /**
     * 中序非递归
     *
     * @param root
     * @return
     */
    public List<Integer> inorderTraversal(TreeNode root) {

        List<Integer> list = new ArrayList<>();

        Stack<TreeNode> stack = new Stack<>();

        TreeNode p = root;
        while (p != null || !stack.isEmpty()) {
            if (p != null) {
                stack.push(p);
                p = p.left;
            } else {
                p = stack.pop();
                list.add(p.val);
                p = p.right;
            }
        }

        return list;
    }

    // 非递归后序遍历
    public List<Integer> postorderTraversal(TreeNode root) {
        List<Integer> res = new ArrayList<>();
        if (root == null) {
            return res;
        }

        Stack<TreeNode> stack = new Stack<>();

        TreeNode p = root;

// 标记最近出栈的节点，用于判断是否是p节点的右孩子，如果是的话，就可以访问p节点
        TreeNode pre = p;

        while (p != null || !stack.isEmpty()) {
            if (p != null) {

                stack.push(p);
                p = p.left;

            } else {
                p = stack.pop();

                if (p.right == null || p.right == pre) {
                    res.add(p.val);
//                    pre = cur;
                    p = null;
                } else {
                    stack.push(p);
                    p = p.right;
                    stack.push(p);
                    p = p.left;
                }
            }
        }

        return res;
    }
}

File: algorithm/src/main/java/com/qibenyu/algorithm/Tree.kt
package com.qibenyu.algorithm

import java.lang.StringBuilder


data class TreeNode(val value: String, var left: TreeNode? = null, var right: TreeNode? = null)

fun TreeNode?.serialized(): String {
    val sb = StringBuilder()
    serialized(this, sb)
    return sb.toString()
}

private fun serialized(treeNode: TreeNode?, sb: StringBuilder) {
    if (treeNode == null) {
        sb.append("#")
        return
    }
    sb.append(treeNode.value)
    serialized(treeNode.left, sb)
    serialized(treeNode.right, sb)
}

fun String.deserialized(): TreeNode? {
    return deserialized(Pointer(0), this)
}

data class Pointer(var index: Int = 0)

private fun deserialized(pointer: Pointer, treeSerialized: String): TreeNode? {

    if (treeSerialized[pointer.index] == '#') {
        pointer.index++
        return null
    }

    val treeNode = TreeNode(treeSerialized[pointer.index].toString())
    pointer.index++
    treeNode.left = deserialized(pointer, treeSerialized)
    treeNode.right = deserialized(pointer, treeSerialized)

    return treeNode
}

fun preOrderTraverse(node: TreeNode) {
    var root = node

}



